Split glyph runs early and fix some run splitting bugs

astiob opened the pull request:
The Holy Grail of glyph runs is (almost) here. This wasn’t even hard, so far.

My next PR will be karaoke that makes use of the early-detected runs.

astiob commented:
Gah, I messed up breaking runs on karaoke while splitting up the PRs. I'll fix this.

astiob commented:
Karaoke mishap fixed.

Added a commit replacing `emblevel % 2` by `FRIBIDI_LEVEL_IS_RTL(emblevel)`: the less we have to know about it, the better.

Ready for review.

MrSmile commented on libass/ass_render.c:
```diff
@@ -1850,40 +1851,42 @@ fix_glyph_scaling(ASS_Renderer *priv, GlyphInfo *glyph)
-        ((last->flags ^ info->flags) & ~DECO_ROTATE);
+    for (int i = 1; i < render_priv->text_info.length; i++) {
+        GlyphInfo *info = render_priv->text_info.glyphs + i;
+        GlyphInfo *last = render_priv->text_info.glyphs + (i - 1);
```
I think that `last` is misleading here, shouldn't it be `prev` or something?

astiob replied:
Well, “last” can also mean “prev”, and here I tried to keep as much code unmodified as I could, barring the indentation change. We use `last...` for “most recent X” in a few places, although some of them skip some elements so their “most recent” is not always the immediately preceding element. But e. g. in `ass_shaper_find_runs` it actually is.

MrSmile reviewed (approved):
LGTM.

On a side note, `GlyphInfo` currently is a huge mess. At some point in time we should merge that bunch of boolean flags together into one (and also eliminate use of pointers as boolean flags).

astiob commented:
Pushed a two-line change to drop the `effect` field from `GlyphInfo` as it’s no longer used.

astiob commented:
Pushed a few changes:

1. “don’t mingle run ID and bidi embedding level” was somehow missing the hunk that would actually remove the addition. Now it is in place.

2. `glyphs[0].starts_new_run` is now true. I’m conflicted on whether I should rename the field to `starts_run`.

3. Shaping can mark the `starts_new_run` glyph infos with `skip`, so to ensure the run breaks are not missed when combining bitmaps (and in future in other places), it now sets `starts_new_run` on the next nonskipped glyph info.

4. A couple extra fixes to ensure drawings are absolutely not shaped by HarfBuzz.

MrSmile commented on libass/ass_shaper.c:
```diff
@@ -956,6 +956,16 @@ int ass_shaper_shape(ASS_Shaper *shaper, TextInfo *text_info)
 #endif
 
+    // Make sure we don't lose any run breaks due to skipped glyphs
+    for (i = 0; i < text_info->length; i++) {
```
It should be possible to rearrange checks without nested loops. Something like
```C
bool new_run = true;
for (int i = 0; i < text_info->length; i++) {
    if (glyphs[i].starts_new_run)
        new_run = true;
    if (glyphs[i].skip)
        continue;
    glyphs[i].starts_new_run = new_run;
    new_run = false;
}
```

astiob replied:
Is that better though? It seems harder to understand.

MrSmile replied:
For me it's instead simpler to understand than nested loops with a common index variable, but feel free to ignore me.
I think that triple copy-paste of condition `i < text_info->length` isn't that great.

MrSmile commented on libass/ass_shaper.c:
```diff
@@ -639,18 +639,20 @@ static void shape_harfbuzz(ASS_Shaper *shaper, GlyphInfo *glyphs, size_t len)
         // advance in text until end of run
-        while (i < (len - 1) && level == glyphs[i+1].shape_run_id)
+        while (i < (len - 1) && run_id == glyphs[i+1].shape_run_id &&
+                level == shaper->emblevels[i+1])
```
Shouldn't it be `[i + 1]` per libass style?

astiob replied:
Changed both instances to `[i + 1]`.

astiob commented:
Added a commit to fix UB pointer decrement, which was most likely innocent in practice.

astiob commented:
Sorry for wasting a bit of review effort, but… I’ve rolled back this:

> to ensure the run breaks are not missed when combining bitmaps (and in future in other places), [shaping] now sets `starts_new_run` on the next nonskipped glyph info [when it marks the `starts_new_run` glyph infos with `skip`]

I originally thought this would be easier to work with, but now I’ve realized that it makes karaoke unable to distinguish between run-leading-skipped-infos-within-a-single-run and multiple-proper-run-breaks-with-only-skipped-glyphs-in-between.

So now, instead, `starts_new_run` is set only on the first glyph of each run, even if that glyph is skipped. The bitmap combining loop remembers this flag across skipped glyphs, just like it used to remember the `linebreak` flag before this PR.

Further, to avoid adding new bad code only to fix it in the upcoming karaoke PR, I’ve redone the way line breaks affect `starts_new_run`. Previously in this PR, `wrap_lines_smart` set `starts_new_run` on the first nonskipped glyph of the new line after each line break. This was an arbitrary choice I made before considering how this interacts with karaoke, although a pretty good one. But for karaoke compatibility, `starts_new_run` must be set on the first _non-trimmed-whitespace_ glyph of the line (which may still be skipped due to shaping), _including the first line._ So I’ve moved the code from `wrap_lines_smart` to `trim_whitespace` and adjusted accordingly.

Only the “Split glyph runs early” commit has changed, but please re-review it.

Finally: doesn’t the last loop in `wrap_lines_smart` write past the end of a buffer if the last line is entirely whitespace? https://github.com/libass/libass/blob/0aec12b3576e93f8387992de73bca99ab6e50d2f/libass/ass_render.c#L1760-L1774

MrSmile commented on libass/ass_render.c:
```diff
@@ -1850,40 +1854,46 @@ fix_glyph_scaling(ASS_Renderer *priv, GlyphInfo *glyph)
+    for (int i = 0; i < render_priv->text_info.length; i++) {
+        GlyphInfo *info = render_priv->text_info.glyphs + i;
+        if (!i) {
+            info->starts_new_run = true;
```
Is there a need for that? First run is handled by `bool new_run = true;` later.

astiob replied:
Not for bitmap combining, no. But it may make other future uses of this flag simpler: for one, it makes karaoke code a _tiny_ bit simpler. Also, for reference, karaoke will need an `if (!i)` section here in any case.

Also, the flag will anyway be set on the first glyph by `trim_whitespace` if the first line has no leading whitespace. So it wouldn’t be wise to have any code relying on the _absence_ of this flag for the first glyph, either.

But if you prefer to avoid setting this on the first glyph explicitly, I don’t mind removing this line and adding a `|| !i` in karaoke later. We’ll need to keep in mind that the flag _may or may not_ be set on the first glyph, then.

MrSmile replied:
I think it's better to keep `i = 0` case in this function but outside of this loop.

astiob replied:
I missed the fact that `text_info.length` is guaranteed to be nonzero. (This is checked just before the call to `split_style_runs`.)

I’ve now moved this assignment out of the loop.

astiob commented:
I’ve largely finished my karaoke changes (some of which build on top of this), and I think I’m happy with this PR as it is now.

MrSmile commented:
I don't see any more problems, so if there are no issues with test samples you can merge that.
